webapi/src/main/scala/org/knora/webapi/responders/v1/OntologyResponderV1.scala

```diff
@@ -579,90 +568,83 @@ final case class OntologyResponderV1Live(
-                  predicateIri = OntologyConstants.Rdfs.Comment,
-                  preferredLangs = Some(userProfile.lang, appConfig.fallbackLanguage)
-                ),
-                vocabulary = entityInfo.ontologyIri,
-                valuetype_id = entityInfo
-                  .getPredicateObject(OntologyConstants.KnoraBase.ObjectClassConstraint)
-                  .getOrElse(
-                    throw InconsistentRepositoryDataException(
-                      s"Property $propertyIri has no knora-base:objectClassConstraint"
-                    )
+        propertyDefinitions <-
+          ZIO
+            .collectAll(propertyInfoMap.map { case (propertyIri: IRI, entityInfo: PropertyInfoV1) =>
+              if (entityInfo.isLinkProp) {
+                // It is a linking prop: its valuetype_id is knora-base:LinkValue.
+                // It is restricted to the resource class that is given for knora-base:objectClassConstraint
+                // for the given property which goes in the attributes that will be read by the GUI.
+
+                for {
+                  attributeResType <-
+                    ZIO
+                      .fromOption(entityInfo.getPredicateObject(OntologyConstants.KnoraBase.ObjectClassConstraint))
+                      .orElseFail(noClassConstraintException(propertyIri))
+                } yield PropertyDefinitionInNamedGraphV1(
+                  id = propertyIri,
+                  name = propertyIri,
+                  label = entityInfo.getPredicateObject(
+                    predicateIri = OntologyConstants.Rdfs.Label,
+                    preferredLangs = Some(userProfile.lang, appConfig.fallbackLanguage)
                   ),
-                attributes = valueUtilV1.makeAttributeString(
-                  entityInfo.getPredicateStringObjectsWithoutLang(SalsahGui.GuiAttribute)
-                ),
-                gui_name = entityInfo
-                  .getPredicateObject(SalsahGui.GuiElementProp)
-                  .map(iri => SalsahGuiConversions.iri2SalsahGuiElement(iri))
-              )
-            }
-          }.toVector
+                  description = entityInfo.getPredicateObject(
+                    predicateIri = OntologyConstants.Rdfs.Comment,
+                    preferredLangs = Some(userProfile.lang, appConfig.fallbackLanguage)
+                  ),
+                  vocabulary = entityInfo.ontologyIri,
+                  valuetype_id = OntologyConstants.KnoraBase.LinkValue,
+                  attributes = valueUtilV1.makeAttributeString(
+                    entityInfo
+                      .getPredicateStringObjectsWithoutLang(SalsahGui.GuiAttribute) + valueUtilV1
+                      .makeAttributeRestype(attributeResType)
+                  ),
+                  gui_name = entityInfo
+                    .getPredicateObject(SalsahGui.GuiElementProp)
+                    .map(iri => SalsahGuiConversions.iri2SalsahGuiElement(iri))
+                )
```

Comment:
I am not big fan of nesting complex code inside a constructor call.
Code like `vocabulary = entityInfo.ontologyIri` could occasionally be ok. But multiline expressions like for the `description`, `attributes` and such make my eyes twitch as it makes it unnecessarily hard to read.
Would it be possible to extract this as variables?

Reply:
The same problem exists in other parts of the code you have changed.

Reply:
I'm not sure about that, sometimes when values are extracted but spread around the scopes it's harder to read for me, but I will your suggestion.